filters/auth/oidc.go

```diff
@@ -0,0 +1,561 @@
+	"golang.org/x/oauth2"
+)
+
+const (
+	OidcUserInfoName  = "oauthOidcUserInfo"
+	OidcAnyClaimsName = "oauthOidcAnyClaims"
+	OidcAllClaimsName = "oauthOidcAllClaims"
+
+	oauthOidcCookieName = "skipperOauthOidc"
+	stateValidity       = 1 * time.Minute
+)
+
+type (
+	tokenOidcSpec struct {
+		typ         roleCheckType
+		SecretsFile string
+	}
+
+	tokenOidcFilter struct {
+		typ             roleCheckType
+		config          *oauth2.Config
+		provider        *oidc.Provider
+		verifier        *oidc.IDTokenVerifier
+		claims          []string
+		validity        time.Duration
+		secretsFile     string
+		cookiename      string
+		refreshInterval time.Duration
+		closer          chan int
+		redirectPath    string
+		encrypter       *encrypter
+		authCodeOptions []oauth2.AuthCodeOption
+	}
+
+	userInfoContainer struct {
+		OAuth2Token *oauth2.Token
+		UserInfo    *oidc.UserInfo
+		Subject     string
+	}
+
+	claimsContainer struct {
+		OAuth2Token *oauth2.Token
+		Claims      map[string]interface{}
+		Subject     string
+	}
+)
+
+// NewOAuthOidcUserInfos creates filter spec which tests user info.
+func NewOAuthOidcUserInfos(secretsFile string) filters.Spec {
+	return &tokenOidcSpec{typ: checkOIDCUserInfo, SecretsFile: secretsFile}
+}
+
+// NewOAuthOidcAnyClaims creates a filter spec which verifies that the token
+// has one of the claims specified
+func NewOAuthOidcAnyClaims(secretsFile string) filters.Spec {
+	return &tokenOidcSpec{typ: checkOIDCAnyClaims, SecretsFile: secretsFile}
+}
+
+// NewOAuthOidcAllClaims creates a filter spec which verifies that the token
+// has all the claims specified
+func NewOAuthOidcAllClaims(secretsFile string) filters.Spec {
+	return &tokenOidcSpec{typ: checkOIDCAllClaims, SecretsFile: secretsFile}
+}
+
+// CreateFilter creates an OpenID Connect authorization filter.
+//
+// first arg: a provider, for example "https://accounts.google.com",
+//            which has the path /.well-known/openid-configuration
+//
+// Example:
+//
+//     tokenOidcSpec(""https://identity-provider.example.com", "client-id", "clent-secret", "https://callback-url")  -> "https://internal.example.org";
+func (s *tokenOidcSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) <= 4 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	providerURL, err := url.Parse(sargs[0])
+
+	ctx := context.Background()
+	provider, err := oidc.NewProvider(ctx, providerURL.String())
+	if err != nil {
+		log.Errorf("Failed to create new provider %s: %v", providerURL, err)
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	if err != nil {
+		log.Errorf("Failed to create ciphersuite: %v", err)
+		return nil, filters.ErrInvalidFilterParameters
+	}
+	h := sha256.New()
+	for _, s := range sargs {
+		h.Write([]byte(s))
+	}
+	byteSlice := h.Sum(nil)
+	sargsHash := fmt.Sprintf("%x", byteSlice)[:8]
+	generatedCookieName := oauthOidcCookieName + sargsHash
```

Comment:
this means SSO across similar configured filters :)

Reply:
The different filters will have different client IDs and will generate different hashes.